File: docs/test-cases/authentication-failures.go
// Test case: authentication-failures (A07:2025)
package main

import (
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"math/rand"
	"net/http"

	"github.com/golang-jwt/jwt/v5"
)

var jwtSecret = []byte("secret")

// BUG: MD5 used for password hashing (fast, broken, unsalted)
func hashPassword(password string) string {
	sum := md5.Sum([]byte(password))
	return hex.EncodeToString(sum[:])
}

func loginHandler(w http.ResponseWriter, r *http.Request) {
	username := r.FormValue("username")
	password := r.FormValue("password")
	stored := lookupHash(username)
	candidate := hashPassword(password)

	// BUG: timing-unsafe string comparison via == leaks hash bytes
	if stored != candidate {
		http.Error(w, "denied", http.StatusUnauthorized)
		return
	}
	// BUG: session ID from math/rand.Intn — predictable, not crypto/rand
	sessionID := rand.Intn(1 << 30)
	fmt.Fprintf(w, "session=%d", sessionID)
}

func verifyToken(tokenStr string) (*jwt.Token, error) {
	// BUG: keyfunc returns secret without checking token.Method — accepts alg:none
	return jwt.Parse(tokenStr, func(t *jwt.Token) (interface{}, error) {
		return jwtSecret, nil
	})
}

func lookupHash(username string) string { return "" }

func main() { http.HandleFunc("/login", loginHandler); http.ListenAndServe(":8080", nil) }

File: docs/test-cases/authentication-failures.java
// Test case: authentication-failures (A07:2025)
package com.example.auth;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import java.security.MessageDigest;
import java.util.Random;
import org.springframework.web.bind.annotation.*;

@RestController
public class AuthController {

    private static final String JWT_KEY = "secret";
    private final Random random = new Random();

    // BUG: SHA-1 (unsalted, fast hash) used for password storage
    public String storePassword(String password) throws Exception {
        MessageDigest md = MessageDigest.getInstance("SHA-1");
        byte[] digest = md.digest(password.getBytes());
        StringBuilder sb = new StringBuilder();
        for (byte b : digest) sb.append(String.format("%02x", b));
        return sb.toString();
    }

    @PostMapping("/login")
    public String login(@RequestParam String username, @RequestParam String password) throws Exception {
        String stored = lookupHash(username);
        String candidate = storePassword(password);
        // BUG: String.equals is timing-unsafe — leaks hash bytes via response time
        if (!stored.equals(candidate)) {
            return "denied";
        }
        // BUG: session token from java.util.Random (non-cryptographic PRNG)
        long sessionId = random.nextLong();
        return "session=" + Long.toHexString(sessionId);
    }

    public Claims verifyToken(String token) {
        // BUG: parser does not pin algorithm — accepts alg:none / HS256 confusion
        return Jwts.parser().setSigningKey(JWT_KEY).parseClaimsJws(token).getBody();
    }

    private String lookupHash(String username) { return ""; }
}
